benchmarks/be_loader.erl

```diff
@@ -0,0 +1,260 @@
+-spec(start_link(Id :: atom()) ->
+  {ok, Pid :: pid()} | ignore | {error, Reason :: term()}).
+start_link(Id) ->
+  gen_server:start_link({local, Id}, ?MODULE, [Id], []).
+
+%% @doc Stops the server
+stop(ServerRef) ->
+  gen_server:stop(ServerRef).
+
+%% @doc Loads Boolean Expressions from file
+load(ServerRef, FileName) ->
+  gen_server:call(ServerRef, {load, FileName}, infinity).
+
+%%%===================================================================
+%%% gen_server callbacks
+%%%===================================================================
+
+%% @private
+%% @doc Initializes the server
+-spec(init(Args :: term()) ->
+  {ok, State :: #be_loader_state{}} | {ok, State :: #be_loader_state{}, timeout() | hibernate} |
+  {stop, Reason :: term()} | ignore).
+init([Id]) ->
+  {ok, #be_loader_state{id = Id, load_count = 0}}.
+
+%% @private
+%% @doc Handling call messages
+-spec(handle_call(Request :: term(), From :: {pid(), Tag :: term()},
+    State :: #be_loader_state{}) ->
+  {reply, Reply :: term(), NewState :: #be_loader_state{}} |
+  {reply, Reply :: term(), NewState :: #be_loader_state{}, timeout() | hibernate} |
+  {noreply, NewState :: #be_loader_state{}} |
+  {noreply, NewState :: #be_loader_state{}, timeout() | hibernate} |
+  {stop, Reason :: term(), Reply :: term(), NewState :: #be_loader_state{}} |
+  {stop, Reason :: term(), NewState :: #be_loader_state{}}).
+handle_call({load, FileName}, _From, State = #be_loader_state{
+  id = Id, load_count = LoadCount}) ->
+  case be_bm_utils:file_exists(FileName) of
+    false ->
+      {reply, {error, {file_does_not_exist, FileName}}, State};
+    true ->
+      StrId = atom_to_list(Id),
+      LoadCount1 = LoadCount + 1,
+      StrCount = integer_to_list(LoadCount1),
+      ReaderId = list_to_atom(StrId ++ "_" ++ StrCount ++ "_loader_reader"),
+      case term_reader:start_link(ReaderId, FileName) of
+        {error, _Reason} = Err ->
+          {reply, Err, State};
+        {ok, _ReaderState} ->
+          case term_reader:read(ReaderId) of
+            eof ->
+              term_reader:stop(ReaderId),
+              {reply, {error, be_tree_parameters_not_provided}, State};
+            {error, _Reason} = Err ->
+              term_reader:stop(ReaderId),
+              {reply, Err, State};
+            {ok, Term} ->
+              Domains = [Term],
+              case erl_betree:betree_make(Domains) of
+                {ok, Betree} ->
+                  EvalId = list_to_atom(StrId ++ "_" ++ StrCount ++ "_evaluator"),
+                  Consts = [],
+                  Index = 0,
+                  BeEvaluator = #be_evaluator{
+                    betree = Betree,
+                    consts = Consts,
+                    index = Index
+                  },
+                  {ok, _StateEval} = term_eval:start_link(EvalId, fun add_expr/2, BeEvaluator),
+
+                  StatsId = list_to_atom(StrId ++ "_" ++ StrCount ++ "_stats"),
```

Comment:
Atom leak?

Reply:
@mijkenator noticed the same. I tried to explain here
https://github.com/adgear/erl-be-tree/pull/6#discussion_r1528999088

Reply:
@saleyn I removed atoms generation based on increasing counters.